pkg/abstractions/secret/secret.go

```diff
@@ -0,0 +1,164 @@
+package secret
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/beam-cloud/beta9/pkg/auth"
+	"github.com/beam-cloud/beta9/pkg/repository"
+	pb "github.com/beam-cloud/beta9/proto"
+	"github.com/labstack/echo/v4"
+	"github.com/lib/pq"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type SecretService interface {
+	pb.SecretServiceServer
+	CreateSecret(ctx context.Context, req *pb.CreateSecretRequest) (*pb.CreateSecretResponse, error)
+	GetSecret(ctx context.Context, req *pb.GetSecretRequest) (*pb.GetSecretResponse, error)
+	ListSecrets(ctx context.Context, req *pb.ListSecretsRequest) (*pb.ListSecretsResponse, error)
+	UpdateSecret(ctx context.Context, req *pb.UpdateSecretRequest) (*pb.UpdateSecretResponse, error)
+	DeleteSecret(ctx context.Context, req *pb.DeleteSecretRequest) (*pb.DeleteSecretResponse, error)
+}
+
+type WorkspaceSecretService struct {
+	pb.UnimplementedSecretServiceServer
+	backendRepo repository.BackendRepository
+}
+
+var secretRoutePrefix = "/secret"
+
+func NewSecretService(backendRepo repository.BackendRepository, routeGroup *echo.Group) SecretService {
+	ss := &WorkspaceSecretService{
+		backendRepo: backendRepo,
+	}
+
+	// Register HTTP routes
+	authMiddleware := auth.AuthMiddleware(backendRepo)
+	registerSecretRoutes(routeGroup.Group(secretRoutePrefix, authMiddleware), ss)
+
+	return ss
+}
+
+func handleErrMsg(err error) string {
+	msg := err.Error()
+
+	if err, ok := err.(*pq.Error); ok {
+		if err.Code == "23505" { // Unique violation
```

Comment:
need to clarify this a bit, reading this I would have no clue what this magic number is / why it's needed